internal/httpfile/httpfile.go

```diff
@@ -0,0 +1,429 @@
+package httpfile
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+// Request represents a request made by a File instance.
+// It will usually be mapped into an HTTP request
+// to the object that has been opened.
+type Request struct {
+	Method string
+	Header http.Header
+}
+
+// Response represents a response from a Request.
+// It will usually be created from an HTTP response.
+type Response struct {
+	StatusCode    int
+	Header        http.Header
+	ContentLength int64
+	Body          io.ReadCloser
+}
+
+type Client interface {
+	// Do sends an HTTP request to the file.
+	// The provided header fields are added
+	// to the request, and the given request is
+	// send with the given method (GET or HEAD).
+	Do(req *Request) (*Response, error)
+}
+
+// SupportedResponseStatuses holds the set of HTTP statuses that a
+// response may legimately have.
+var SupportedResponseStatuses = []int{
+	http.StatusOK,
+	http.StatusNotFound,
+	http.StatusPartialContent,
+	http.StatusPreconditionFailed,
+}
+
+var ErrNotFound = errors.New("file not found")
+
+var errChanged = errors.New("file has changed since it was opened")
+
+type File struct {
+	client Client
+	length int64
+	etag   string
+
+	// readAhead holds the number of bytes
+	// to request in advance of any Read.
+	readAhead int64
+
+	// pos holds the current seek position of the reader.
+	pos int64
+
+	// reader0 represents the currently running GET request.
+	reader0 *readerAhead
+
+	// reader1 represents the speculative GET request
+	// one block ahead of reader0. This is never
+	// non-nil when reader0 is nil.
+	reader1 *readerAhead
+}
+
+var supportedReadStatuses = []int{http.StatusOK, http.StatusPartialContent, http.StatusPreconditionFailed}
+
+var errOutOfRange = errors.New("read out of range")
+
+func Open(c Client, readAhead int64) (*File, http.Header, error) {
+	if readAhead == -1 {
+		readAhead = infinity
+	}
+	f := &File{
+		client:    c,
+		length:    -1, // Unknown.
+		readAhead: readAhead,
+	}
+	f.reader0 = f.newReaderAhead(0, readAhead)
```

Comment:
I suspect that if you always did a HEAD request here to get the length and Etag etc. then a number of the special cases in the code below would disappear making things simpler.